### PR TITLE
Constant force

### PR DESCRIPTION
**Pull request by SchoeniPhlippsn**
Reimplemented and rewritten the constant force class  

## Description

Users can now add a constant force and constant torque on particles selected by a filter. 

## Motivation and context

It was a feature of HOOMD v2 and is faster than using the custom force approach

## How has this been tested?

Added pytests and ran some simulations with it.

## Change log

```
Reimplementation of a constant force and constant torque class
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/md/ConstantForceCompute.cc [hunk 1]
```diff
@@ -0,0 +1,219 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#include "ConstantForceCompute.h"
+#include "hoomd/RNGIdentifiers.h"
+#include "hoomd/RandomNumbers.h"
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
```
Not used.

File: hoomd/md/ConstantForceCompute.cc [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+namespace hoomd
+    {
+namespace md
+    {
+/*! \file ConstantForceCompute.cc
+    \brief Contains code for the ConstantForceCompute class
+*/
+
+/*! \param rotation_diff rotational diffusion constant for all particles.
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
/*! \param group Group to apply constant force to.
```

File: hoomd/md/ConstantForceCompute.cc [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    m_f_constantVec.swap(tmp_f_constantVec);
+    TAG_ALLOCATION(m_f_constantVec);
+
+    ArrayHandle<Scalar3> h_f_constantVec(m_f_constantVec,
+                                         access_location::host,
+                                         access_mode::overwrite);
+    for (unsigned int i = 0; i < m_f_constantVec.size(); i++)
+        h_f_constantVec.data[i] = make_scalar3(0.0, 0.0, 0.0);
+
+    // allocate memory for the per-type constant_force storage and initialize them to (0,0,0)
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
    // allocate memory for the per-type constant torque storage and initialize them to (0,0,0)
```

File: hoomd/md/ConstantForceCompute.cc [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+/*! \param rotation_diff rotational diffusion constant for all particles.
+ */
+ConstantForceCompute::ConstantForceCompute(std::shared_ptr<SystemDefinition> sysdef,
+                                           std::shared_ptr<ParticleGroup> group)
+
+    : ForceCompute(sysdef), m_group(group), m_need_rearrange_forces(false)
+    {
+    // allocate memory for the per-type constant_force storage and initialize them to (1.0,0,0)
+    GlobalVector<Scalar3> tmp_f_constantVec(m_pdata->getNTypes(), m_exec_conf);
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
    GlobalVector<Scalar3> tmp_f(m_pdata->getNTypes(), m_exec_conf);
```
This variable name mixes naming conventions. I think `tmp_f` is sufficiently descriptive.

File: hoomd/md/ConstantForceCompute.cc [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    TAG_ALLOCATION(m_f_constantVec);
+
+    ArrayHandle<Scalar3> h_f_constantVec(m_f_constantVec,
+                                         access_location::host,
+                                         access_mode::overwrite);
+    for (unsigned int i = 0; i < m_f_constantVec.size(); i++)
+        h_f_constantVec.data[i] = make_scalar3(0.0, 0.0, 0.0);
+
+    // allocate memory for the per-type constant_force storage and initialize them to (0,0,0)
+    GlobalVector<Scalar3> tmp_t_constantVec(m_pdata->getNTypes(), m_exec_conf);
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
    GlobalVector<Scalar3> tmp_t(m_pdata->getNTypes(), m_exec_conf);
```
Same here.

File: hoomd/md/ConstantForceCompute.cc [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    m_exec_conf->msg->notice(5) << "Destroying ConstantForceCompute" << std::endl;
+    }
+
+void ConstantForceCompute::setConstantForce(const std::string& type_name, pybind11::tuple v)
+    {
+    unsigned int typ = this->m_pdata->getTypeByName(type_name);
+
+    if (pybind11::len(v) != 3)
+        {
+        throw std::invalid_argument("gamma_r values must be 3-tuples");
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
        throw std::invalid_argument("force values must be 3-tuples");
```

File: hoomd/md/ConstantForceCompute.cc [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+    // check for user errors
+    if (typ >= m_pdata->getNTypes())
+        {
+        throw std::invalid_argument("Type does not exist");
+        }
+
+    Scalar3 f_constantVec;
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
    Scalar3 force;
```
Naming conventions.

File: hoomd/md/ConstantForceCompute.cc [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+pybind11::tuple ConstantForceCompute::getConstantForce(const std::string& type_name)
+    {
+    pybind11::list v;
+    unsigned int typ = this->m_pdata->getTypeByName(type_name);
+
+    ArrayHandle<Scalar3> h_f_constantVec(m_f_constantVec, access_location::host, access_mode::read);
+
+    Scalar3 f_constantVec = h_f_constantVec.data[typ];
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
    Scalar3 force = h_f_constantVec.data[typ];
```
Naming conventions.

File: hoomd/md/ConstantForceCompute.cc [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    return pybind11::tuple(v);
+    }
+
+void ConstantForceCompute::setConstantTorque(const std::string& type_name, pybind11::tuple v)
+    {
+    unsigned int typ = this->m_pdata->getTypeByName(type_name);
+
+    if (pybind11::len(v) != 3)
+        {
+        throw std::invalid_argument("gamma_r values must be 3-tuples");
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
        throw std::invalid_argument("torque values must be 3-tuples");
```

File: hoomd/md/ConstantForceCompute.cc [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+    // check for user errors
+    if (typ >= m_pdata->getNTypes())
+        {
+        throw std::invalid_argument("Type does not exist");
+        }
+
+    Scalar3 t_constantVec;
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
    Scalar3 torque;
```

File: hoomd/md/ConstantForceCompute.cc [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+    m_need_rearrange_forces = true;
+    }
+
+pybind11::tuple ConstantForceCompute::getConstantTorque(const std::string& type_name)
+    {
+    pybind11::list v;
+    unsigned int typ = this->m_pdata->getTypeByName(type_name);
+
+    ArrayHandle<Scalar3> h_t_constantVec(m_t_constantVec, access_location::host, access_mode::read);
+    Scalar3 t_constantVec = h_t_constantVec.data[typ];
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.cc]:
```suggestion
    Scalar3 torque = h_t_constantVec.data[typ];
```

File: hoomd/md/ConstantForceCompute.h [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    protected:
+    //! Actually compute the forces
+    virtual void computeForces(uint64_t timestep);
+
+    //! Set forces for particles
+    virtual void setForces();
+
+    std::shared_ptr<ParticleGroup> m_group; //!< Group of particles on which this force is applied
+    GlobalVector<Scalar3>
+        m_f_constantVec; //! constant force unit vectors and magnitudes for each particle type
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.h]:
```suggestion
        m_constant_force; //! constant force vectors for each particle type
```
Similarly, all handles to this should be `h_constant_force` or `d_constant_force`.

File: hoomd/md/ConstantForceCompute.h [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+    GlobalVector<Scalar3>
+        m_t_constantVec; //! constant torque unit vectors and magnitudes for each particle type
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.h]:
```suggestion
        m_constant_torque; //! constant torque for each particle type
```
Similarly, all handles to this should be `h_constant_torque` or `d_constant_torque`.

File: hoomd/md/ConstantForceCompute.h [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+    bool m_need_rearrange_forces; //!< True if forces need to be rearranged
```
Comment (joaander) [inline hoomd/md/ConstantForceCompute.h]:
```suggestion
    bool m_parameters_updated; //!< Set to true when the user changes parameters.
```
The old variable name doesn't reflect the current usage.

File: hoomd/md/ConstantForceComputeGPU.cc [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        for (unsigned int i = 0; i < type; i++)
+            {
+            f_constantVec.data[i] = old_f_constantVec.data[i];
+
+            t_constantVec.data[i] = old_t_constantVec.data[i];
+            }
+        }
+
+    m_f_constantVec.swap(tmp_f_constantVec);
+    m_t_constantVec.swap(tmp_t_constantVec);
```
Comment (joaander) [inline hoomd/md/ConstantForceComputeGPU.cc]:
```suggestion
```
Why is this here? The base class initializes these member variables, so there is no need to reinitialize them.

File: hoomd/md/ConstantForceComputeGPU.cc [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+                                                  d_f_constVec.data,
+                                                  d_t_constVec.data,
+                                                  N,
+                                                  m_tuner->getParam()[0]);
+
+    if (m_exec_conf->isCUDAErrorCheckingEnabled())
+        CHECK_CUDA_ERROR();
+
+    m_tuner->end();
+    }
```
Comment (joaander) [inline hoomd/md/ConstantForceComputeGPU.cc:84]: This fails to set `m_need_rearrange_forces = false;`, so `setForces` will be called every time step. Better to set `m_need_rearrange_forces = false;` in the base class `computeForces` method.

File: hoomd/md/force.py [hunk 17]
```diff
@@ -521,3 +521,83 @@ def _set_cpp_obj(self):
...
+    by the filter:
+
+    Examples::
+
+        all = hoomd.filter.All()
+        constant = hoomd.md.force.Constant(
+            filter=hoomd.filter.All()
+            )
+        constant.constant_force['A','B'] = (1,0,0)
+        constant.constant_torque['A','B'] = (0,0,0)
```
Comment (joaander) [inline hoomd/md/force.py]:
```suggestion
        constant.constant_force['A'] = (1,0,0)
        constant.constant_torque['A'] = (0,0,0)
```

File: hoomd/md/force.py [hunk 18]
```diff
@@ -521,3 +521,83 @@ def _set_cpp_obj(self):
...
+    Args:
+        filter (`hoomd.filter`): Subset of particles on which to
+            apply constant forces.
+
+    `Constant` computes an constant force and torque on all particles selected
+    by the filter:
+
+    Examples::
+
+        all = hoomd.filter.All()
```
Comment (joaander) [inline hoomd/md/force.py]:
```suggestion
```

File: hoomd/md/force.py [hunk 19]
```diff
@@ -521,3 +521,83 @@ def _set_cpp_obj(self):
...
+
+class Constant(Force):
+    r"""Constant force.
+
+    Args:
+        filter (`hoomd.filter`): Subset of particles on which to
+            apply constant forces.
+
+    `Constant` computes an constant force and torque on all particles selected
+    by the filter:
```
Comment (joaander) [inline hoomd/md/force.py]:
```suggestion
    `Constant` applies a type dependent constant force and torque on all particles selected
    by the filter:
```

File: hoomd/md/pytest/test_constant_force.py [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+    sim.run(0)
+
+    assert constant.constant_force['A'] == (0.0, 0.0, 0.0)
+    assert constant.constant_torque['A'] == (0.0, 0.0, 0.0)
+
+    constant.constant_force['A'] = (0.5, 0.0, 0.0)
+    assert constant.constant_force['A'] == (0.5, 0.0, 0.0)
+    constant.constant_force['A'] = (0.0, 0.0, 1.0)
+    assert constant.constant_force['A'] == (0.0, 0.0, 1.0)
+
```
Comment (joaander) [inline hoomd/md/pytest/test_constant_force.py:98]: Check the values of `constant.force` and `constant.torque` as well.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks for this, it will be very useful. I have a number of requested changes, most only on variable naming conventions. There is one performance bug I noticed.

Do you have a strong use-case that justifies the need for type-specific parameters in addition to the filter? When does this work better than multiple `force.Constant` instances with different filters and only single fixed force and torque values? I ask because the current behavior is somewhat counterintuitive and will take some additional documentation to explain fully. Users will be confused why there are two ways to select which particles get which forces.

File: hoomd/md/pytest/test_constant_force.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+    sim.run(0)
+
+    assert constant.constant_force['A'] == (0.0, 0.0, 0.0)
+    assert constant.constant_torque['A'] == (0.0, 0.0, 0.0)
+
+    constant.constant_force['A'] = (0.5, 0.0, 0.0)
+    assert constant.constant_force['A'] == (0.5, 0.0, 0.0)
+    constant.constant_torque['A'] = (0.0, 0.0, 1.0)
+    assert constant.constant_torque['A'] == (0.0, 0.0, 1.0)
+
```
Comment (joaander) [inline hoomd/md/pytest/test_constant_force.py:46]:
```suggestion
assert numpy.assert_array_equal(constant.forces, [[0.5, 0.0, 0.0], [0.5, 0.0, 0.0]])
assert numpy.assert_array_equal(constant.torques, [[0.0, 0.0, 1.0], [0.0, 0.0, 1.0]])
```

File: hoomd/md/pytest/test_constant_force.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
+# Copyright (c) 2009-2022 The Regents of the University of Michigan.
+# Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+import hoomd
+from hoomd.conftest import pickling_check, autotuned_kernel_parameter_check
+
```
Comment (joaander) [inline hoomd/md/pytest/test_constant_force.py:7]:
```suggestion
import numpy
```

File: hoomd/md/force.py [hunk 23]
```diff
@@ -521,3 +521,82 @@ def _set_cpp_obj(self):
...
+
+class Constant(Force):
+    r"""Constant force.
+
+    Args:
+        filter (`hoomd.filter`): Subset of particles on which to
+            apply constant forces.
+
+    `Constant` applies a type dependent constant force and torque on all
+    particles selected by the filter:
```
Comment (joaander) [inline hoomd/md/force.py]:
```suggestion
    particles selected by the filter. `Constant` sets the force and torque to 
    ``(0,0,0)`` for particles not selected by the filter.

```

File: hoomd/md/ConstantForceComputeGPU.cu [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+                                                 const Scalar3* d_t_const,
+                                                 const unsigned int N,
+                                                 unsigned int block_size)
+    {
+    // setup the grid to run the kernel
+    dim3 grid(group_size / block_size + 1, 1, 1);
+    dim3 threads(block_size, 1, 1);
+
+    // run the kernel
+    hipMemset(d_force, 0, sizeof(Scalar4) * N);
```
Comment (joaander) [inline hoomd/md/ConstantForceComputeGPU.cu:73]:
```suggestion
    hipMemset(d_force, 0, sizeof(Scalar4) * N);
    hipMemset(d_torque, 0, sizeof(Scalar4) * N);
```

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks, this is a big improvement!

In addition to the in-line suggested `numpy.assert_array_equal(constant.forces ...`, please add tests with 1) multiple particle types and 2) filters that are not `All` to ensure that the output in `forces` and `torques` is as expected.

File: hoomd/md/force.py [hunk 25]
```diff
@@ -521,3 +521,83 @@ def _set_cpp_obj(self):
...
+
+    Attributes:
+        filter (`hoomd.filter`): Subset of particles on which to
+            apply constant forces.
+
+    .. py:attribute:: constant_force
+
+        Constant force vector in the global reference frame of the system
+        :math:`[\mathrm{force}]`.  It is defined per particle type and stays
+        constant during the simulation.
```
Comment (joaander) [inline hoomd/md/force.py]:
```suggestion
        :math:`[\mathrm{force}]`.  It is defined per particle type and 
        defaults to (0.0, 0.0, 0.0) for all types.
```
It is not necessarily constant, the user can change it. Also, document the default value.

File: hoomd/md/force.py [hunk 26]
```diff
@@ -521,3 +521,83 @@ def _set_cpp_obj(self):
...
+
+        Type: `TypeParameter` [``particle_type``, `tuple` [`float`, `float`,
+        `float`]]
+
+    .. py:attribute:: constant_torque
+
+        Constant torque vector in the global reference frame of the system
+        :math:`[\mathrm{force} \cdot \mathrm{length}]`. It is defined per
+        particle type and stays constant during the simulation.
```
Comment (joaander) [inline hoomd/md/force.py]:
```suggestion
        particle type and defaults to (0.0, 0.0, 0.0) for all types.
```

File: hoomd/md/pytest/test_constant_force.py [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+    sim = simulation_factory(two_particle_snapshot_factory(dimensions=3, d=8))
+    integrator = hoomd.md.Integrator(.05)
+    integrator.methods.append(
+        hoomd.md.methods.Langevin(hoomd.filter.All(), kT=0))
+    integrator.forces.append(constant)
+    sim.operations.integrator = integrator
+    sim.run(0)
+
+    def activate_kernel():
+        constant.constant_force['A'] = (1.0, 2.0, 3.0)
```
Comment (joaander) [inline hoomd/md/pytest/test_constant_force.py:112]: I pushed this change to trigger the autotuner to run on every step of the test. Previously, the kernel only executed once.

File: hoomd/md/pytest/test_constant_force.py [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+    sim.run(1)
+
+    if sim.device.communicator.rank == 0:
+        numpy.testing.assert_array_equal(constant.forces,
+                                         [[0.0, 0.0, 0.0], [0.5, 0.0, 0.0]])
+        numpy.testing.assert_array_equal(constant.torques,
+                                         [[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]])
+
+
+def test_types(simulation_factory, two_particle_snapshot_factory):
```
Comment (joaander) [inline hoomd/md/pytest/test_constant_force.py:64]: I added this test to ensure that forces and torques are correctly set by type.

File: hoomd/md/pytest/test_constant_force.py [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+    assert constant.constant_torque['A'] == (0.0, 0.0, 0.0)
+
+    constant.constant_force['A'] = (0.5, 0.0, 0.0)
+    assert constant.constant_force['A'] == (0.5, 0.0, 0.0)
+    constant.constant_force['A'] = (0.0, 0.0, 1.0)
+    assert constant.constant_force['A'] == (0.0, 0.0, 1.0)
+
+
+def test_attach_and_filter(simulation_factory, two_particle_snapshot_factory):
+    constant = hoomd.md.force.Constant(filter=hoomd.filter.Type(['A']))
```
Comment (joaander) [inline hoomd/md/pytest/test_constant_force.py:22]: I modified this test to ensure that the filter is applied as documented.

**Review by joaander — APPROVED** [review]
Thanks! I pushed some changes to address the testing issues.

**Review by shihkual — APPROVED** [review]
Looks good to me.

**Pull request merged**